pkg/reconcileaction/reconcileaction.go

```diff
@@ -0,0 +1,47 @@
+package reconcileaction
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// ReconcileAction is an action that reconciles the system state. It has a list
+// of prerequisite actions that must be true in order for the action to be
+// invoked.
+type ReconcileAction struct {
+	Name          string
+	prereqs       []*ReconcileAction
+	action        func(reconcile.Request) (corev1.ConditionStatus, error)
+	lastCondition *corev1.ConditionStatus
+	lastError     error
+}
+
+// Execute or return a previously cached result of the ReconcileAction, checking prereqs first
+func (ra *ReconcileAction) Execute(request reconcile.Request) (corev1.ConditionStatus, error) {
+	// If we have executed before, return the cached result
+	if ra.lastCondition != nil {
+		return *ra.lastCondition, ra.lastError
+	}
+
+	// Walk through the prereqs; stop and return corev1.ConditionUnknown if a prereq doesn't return corev1.ConditionTrue
+	for _, prereq := range ra.prereqs {
+		cond, err := prereq.Execute(request)
+		if err != nil || cond != corev1.ConditionTrue {
+			unknown := corev1.ConditionUnknown
```

Comment:
Why not return the condition from the failed prereq?

Reply:
The state of the action is unknown... We only know the result of the prereq.
Consider:
- action = ensure devices are registered w/ gd2
- prereq = gd2 node is up

If the node is down or indeterminate, we don't actually know if the devices are registered or not.

Reply:
But in this case there's no distinction between false and unknown, the only part that matters is "not true". Is there need for any conditions other than true and not true?

Reply:
Okay, legit.